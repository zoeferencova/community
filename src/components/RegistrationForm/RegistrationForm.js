import React, { Component } from "react";
import { withRouter } from "react-router-dom";
import { ButtonLight, ButtonDark, Input, Label, Error } from "../Utils/Utils";
import styles from "../LoginForm/LoginForm.module.css"

import AuthApiService from "../../services/auth-api-service";

class RegistrationForm extends Component {    
    state = { error: null };
    
    handleSubmit = e => {
        e.preventDefault();
        const { first_name, email, password, confirm_password } = e.target;
        password.value !== confirm_password.value
            ? this.setState({ error: "Passwords do not match" })
            : AuthApiService.postUser({
                first_name: first_name.value,
                email: email.value,
                password: password.value
            })
                .then(user => {
                    first_name.value = "";
                    email.value = "";
                    password.value = "";
                    confirm_password.value = "";
                    this.props.history.push("/login");
                })
                .catch(res => 
                    this.setState({ error: res.error })
                )
    }
    
    render() {
        return (
            <form onSubmit={e => this.handleSubmit(e)}>
                {this.state.error && <Error message={this.state.error} />}                
                <Label htmlFor="first_name">First name</Label>
                <Input required type="text" name="first_name" id="first_name" />
                <Label htmlFor="email">Email</Label>
                <Input required type="email" name="email" id="email" />
                <Label htmlFor="password">Password</Label>
                <Input required type="password" name="password" id="password" />
                <Label htmlFor="confirm_password">Confirm Password</Label>
                <Input required type="password" name="confirm_password" id="confirm_password" />
                <div className={styles.buttonContainer}>
                    <ButtonLight type="button" onClick={() => this.props.history.push("/")}>Cancel</ButtonLight>
                    <ButtonDark type="submit">Create Account</ButtonDark>
                </div>
                
            </form>
        )
    }
}

export default withRouter(RegistrationForm)