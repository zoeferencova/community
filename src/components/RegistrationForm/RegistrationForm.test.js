import React from "react";
import { mount } from "enzyme";
import RegistrationForm from "./RegistrationForm";
import { BrowserRouter } from "react-router-dom";

describe("RegistrationForm", () => {
    it("renders RegistrationForm component", () => {
        const mockProps = {
            loading: false,
            error: null,
            register: jest.fn()
        }

        const wrapper = mount(
            <BrowserRouter>
                <RegistrationForm {...mockProps} />
            </BrowserRouter>
        );

        expect(wrapper).toMatchSnapshot();
    });

    it("renders an error message if there is a form error", () => {
        let mockProps, wrapper;
        
        mockProps = {
            loading: false,
            error: null,
            register: jest.fn()
        }

        wrapper = mount(
            <BrowserRouter>
                <RegistrationForm {...mockProps} />
            </BrowserRouter>
        );

        expect(wrapper.find("Error").length).toBe(0);

        mockProps = {
            loading: false,
            error: "error",
            register: jest.fn()
        }

        wrapper = mount(
            <BrowserRouter>
                <RegistrationForm {...mockProps} />
            </BrowserRouter>
        );

        expect(wrapper.find("Error").length).toBe(1);
        expect(wrapper.find("Error").text()).toEqual("error");
    });

    it("renders the loading icon when the form submission is loading", () => {
        let wrapper, mockProps;

        mockProps = {
            loading: true,
            error: null,
            register: jest.fn()
        }
        
        wrapper = mount(
            <BrowserRouter>
                <RegistrationForm { ...mockProps } />
            </BrowserRouter>
        );

        expect(wrapper.find("button[type='submit']").html()).toContain("<i class=\"fa fa-spinner fa-spin spinner\"></i>");

        mockProps = {
            loading: false,
            error: null,
            register: jest.fn()
        }

        wrapper = mount(
            <BrowserRouter>
                <RegistrationForm { ...mockProps } />
            </BrowserRouter>
        );
        
        expect(wrapper.find("button[type='submit']").text()).toEqual("Create account");
    });

    it("correctly fires logIn callback on form submission", () => {        
        const mockValues = {
            first_name: "Test",
            email: "test@test.com", 
            password: "Password1!",
            confirm_password: "Password1!"
        }

        const mockProps = {
            loading: false,
            error: null,
            register: jest.fn()
        }

        const wrapper = mount(
            <BrowserRouter>
                <RegistrationForm {...mockProps} />
            </BrowserRouter>
        );

        wrapper.find("RegistrationForm").setState(mockValues)
        wrapper.find("form").simulate("submit");

        expect(mockProps.register).toHaveBeenCalledTimes(1);
        expect(mockProps.register).toHaveBeenCalledWith(mockValues.first_name, mockValues.email, mockValues.password, mockValues.confirm_password);
    })
});