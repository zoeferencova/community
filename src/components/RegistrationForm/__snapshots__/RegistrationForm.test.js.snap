// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegistrationForm renders RegistrationForm component 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <withRouter(RegistrationForm)
      error={null}
      loading={false}
      register={[MockFunction]}
    >
      <RegistrationForm
        error={null}
        history={
          Object {
            "action": "POP",
            "block": [Function],
            "createHref": [Function],
            "go": [Function],
            "goBack": [Function],
            "goForward": [Function],
            "length": 1,
            "listen": [Function],
            "location": Object {
              "hash": "",
              "pathname": "/",
              "search": "",
              "state": undefined,
            },
            "push": [Function],
            "replace": [Function],
          }
        }
        loading={false}
        location={
          Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          }
        }
        match={
          Object {
            "isExact": true,
            "params": Object {},
            "path": "/",
            "url": "/",
          }
        }
        register={[MockFunction]}
      >
        <form
          onSubmit={[Function]}
        >
          <Label
            htmlFor="first_name"
          >
            <label
              className="label "
              htmlFor="first_name"
            >
              First name
            </label>
          </Label>
          <Input
            id="first_name"
            name="first_name"
            onChange={[Function]}
            required={true}
            type="text"
            value=""
          >
            <input
              className="input "
              id="first_name"
              name="first_name"
              onChange={[Function]}
              required={true}
              type="text"
              value=""
            />
          </Input>
          <Label
            htmlFor="email"
          >
            <label
              className="label "
              htmlFor="email"
            >
              Email
            </label>
          </Label>
          <Input
            id="email"
            name="email"
            onChange={[Function]}
            required={true}
            type="email"
            value=""
          >
            <input
              className="input "
              id="email"
              name="email"
              onChange={[Function]}
              required={true}
              type="email"
              value=""
            />
          </Input>
          <Label
            htmlFor="password"
          >
            <label
              className="label "
              htmlFor="password"
            >
              Password
            </label>
          </Label>
          <Input
            id="password"
            name="password"
            onChange={[Function]}
            required={true}
            type="password"
            value=""
          >
            <input
              className="input "
              id="password"
              name="password"
              onChange={[Function]}
              required={true}
              type="password"
              value=""
            />
          </Input>
          <Label
            htmlFor="confirm_password"
          >
            <label
              className="label "
              htmlFor="confirm_password"
            >
              Confirm Password
            </label>
          </Label>
          <Input
            id="confirm_password"
            name="confirm_password"
            onChange={[Function]}
            required={true}
            type="password"
            value=""
          >
            <input
              className="input "
              id="confirm_password"
              name="confirm_password"
              onChange={[Function]}
              required={true}
              type="password"
              value=""
            />
          </Input>
          <div
            className="buttonContainer"
          >
            <ButtonLight
              onClick={[Function]}
              type="button"
            >
              <button
                className="button "
                onClick={[Function]}
                type="button"
              >
                Cancel
              </button>
            </ButtonLight>
            <ButtonDark
              className="registerButton"
              loading="false"
              type="submit"
            >
              <button
                className="buttonDark registerButton"
                loading="false"
                type="submit"
              >
                Create account
              </button>
            </ButtonDark>
          </div>
        </form>
      </RegistrationForm>
    </withRouter(RegistrationForm)>
  </Router>
</BrowserRouter>
`;
